### PR TITLE
any coin inbox

### LINKED ISSUE #878: Base / Any coin inbox

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
https://streamable.com/9lf1lc

Closes #878

### CODE REVIEW

File: packages/daimo-api/src/contract/ethIndexer.ts [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    await Promise.all([
+      this.batchFetchAndCacheBalances(allAddrs, to),
+      this.batchFetchAndCacheBalances(allAddrs, from - 1),
+    ]);
+
+    const newTransfers = allAddrs
+      .map((addr) => {
+        const balanceAfter = this.cachedBalances.get(
+          addrBlockNumKey(addr, to)
+        )!;
```
Comment (dcposch) [inline packages/daimo-api/src/contract/ethIndexer.ts:120]: cachedBalances will grow by ~3000 entries every block, right? do we ever clear/prune it?

Comment (dcposch) [reply packages/daimo-api/src/contract/ethIndexer.ts:120]: possible alternative: delete the cache, do something like this

```ts
const [before, after] = await Promise.all([
  this.batchFetchBalances(allAddrs, to),
  this.batchFetchBalances(allAddrs, from - 1),
]);
  
for (const i in allAddrs) {
  const balanceBefore = before[i];
  const balanceAfter = after[i];
  // ...
}
```

File: packages/daimo-api/src/contract/ethIndexer.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    const [latestBalance, latestBlock] = this.latestBalance.get(addr) || [
+      0n,
+      0,
+    ];
+
+    console.log(
+      `[ETH] getProposedSwap ${addr}: ${latestBalance} at ${latestBlock}`
+    );
+
+    if (!latestBalance) return []; // todo: ignore dust?
```
Comment (dcposch) [inline packages/daimo-api/src/contract/ethIndexer.ts:175]: recommend `if (latestBalance === 0n) ...` here

File: packages/daimo-api/src/contract/foreignCoinIndexer.ts [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+  async preIndexingInit(pg: Pool) {
+    this.foreignTokens = await fetchTokenList();
+    await retryBackoff(`createFilteredTransfersView`, async () => {
+      await pg.query(`CREATE MATERIALIZED VIEW IF NOT EXISTS filtered_erc20_transfers as (SELECT
+        et.*
+      FROM
+        erc20_transfers et
+        JOIN names n ON n.addr = et.f
+          OR n.addr = et.t);`);
+    });
```
Comment (dcposch) [inline packages/daimo-api/src/contract/foreignCoinIndexer.ts:55]: hmm i worry about having db migrations distributed around the codebase

recommend putting with the others in db.ts

```sql
CREATE MATERIALIZED VIEW IF NOT EXISTS filtered_erc20_transfers
AS SELECT et.*
FROM erc20_transfers et
JOIN names n ON n.addr = et.f OR n.addr = et.t
```

File: packages/daimo-api/src/contract/foreignCoinIndexer.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+  // Used to detect onchain swaps initiated by the user.
+  getForeignTokenReceiveForSwap(
+    addr: Address,
+    txHash: Hex
+  ): ForeignTokenTransfer | null {
+    const log = this.sendsByAddrTxHash.get(addrTxHashKey(addr, txHash));
+    if (log == null) return null;
+
+    const correspondingReceive = this.correspondingReceiveOfSend.get(
+      addrTxHashKey(addr, txHash)
```
Comment (dcposch) [inline packages/daimo-api/src/contract/foreignCoinIndexer.ts:201]: 👀

so this matches the in & out transfers of a swap based on (addr, txhash)?

should work for now, but won't in the general case, for example if we have a bigger bundle that contains two userops from the same Daimo account

open challenge, i guess, for Shovel to get us a more elegant way to do this.

shovel indexing internal contract calls would go a long way (eth transfers,  uniswap swaps, etc)

File: packages/daimo-api/src/network/uniswapClient.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  SwapType,
+} from "@uniswap/smart-order-router";
+import { getDefaultProvider } from "ethers";
+import { Address, Hex, getAddress } from "viem";
+
+import { chainConfig } from "../env";
+
+// From https://docs.uniswap.org/contracts/v3/reference/deployments/base-deployments
+export const UNISWAP_V3_02_ROUTER_ADDRESS = getAddress(
+  "0x2626664c2603336E57B271c5C0b26F421741e481"
```
Comment (dcposch) [inline packages/daimo-api/src/network/uniswapClient.ts:31]: move to chainConfig.ts ?

File: packages/daimo-api/src/network/uniswapClient.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    fromCoin: ForeignCoin,
+    receivedAt: number,
+    fromAcc: EAccount
+  ): Promise<ProposedSwap | null> {
+    if (!this.router) return null;
+
+    const options: SwapOptionsSwapRouter02 = {
+      recipient: addr,
+      slippageTolerance: new Percent(50, 10_000), // 50 bips
+      deadline: Math.floor(now() + 1800),
```
Comment (dcposch) [inline packages/daimo-api/src/network/uniswapClient.ts:91]: 30 minutes seems like a long time?

File: packages/daimo-api/src/network/uniswapClient.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    );
+
+    if (!route || !route.methodParameters) {
+      console.log(
+        `[UNISWAP] no route found for ${addr} ${JSON.stringify(
+          fromCoin
+        )}, amount ${fromAmount}`
+      );
+      return null;
+    }
```
Comment (dcposch) [inline packages/daimo-api/src/network/uniswapClient.ts:114]: log the route here

File: packages/daimo-api/src/server/coinList.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+};
+
+// Exclude native ETH
+export type ForeignToken = ForeignCoin & { token: Address };
+
+const customOverrides: Record<Address, ForeignToken> = {
+  "0xd9aAEc86B65D86f6A7B5B1b0c42FFA531710b6CA": {
+    token: "0xd9aAEc86B65D86f6A7B5B1b0c42FFA531710b6CA",
+    fullName: "Bridged USD Coin", // USDBC has a bad name on CoinGecko
+    symbol: "USDBC",
```
Comment (dcposch) [inline packages/daimo-api/src/server/coinList.ts:25]: USDbC

File: packages/daimo-api/src/server/coinList.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    decimals: 6,
+    logoURI: `${daimoDomainAddress}/assets/foreign-coin-logos/USDBC.png`, // CoinGecko logo is fugly
```
Comment (dcposch) [inline packages/daimo-api/src/server/coinList.ts:27]: 😆

File: packages/daimo-api/src/server/pushNotifier.ts [hunk 10]
```diff
@@ -73,6 +87,26 @@ export class PushNotifier {
...
+
+  private handleEthTransfers = async (logs: ETHTransfer[]) => {
+    console.log(`[PUSH] got ${logs.length} foreign coin transfers`);
+    const messages = (
+      await Promise.all(
+        logs.map((log) => this.getPushMessagesFromEthTransfer(log))
+      )
+    ).flat();
+    this.maybeSendNotifications(messages);
+  };
```
Comment (dcposch) [inline packages/daimo-api/src/server/pushNotifier.ts:108]: 🚀

File: packages/daimo-api/src/server/pushNotifier.ts [hunk 11]
```diff
@@ -217,21 +243,30 @@ export class PushNotifier {
...
-      body = `Your ${dollars} ${tokenSymbol} request was fulfilled`;
-    } else if (value < 0) {
-      body = `You sent ${dollars} ${tokenSymbol} to ${otherStr}`;
-    } else {
-      body = `You received ${dollars} ${tokenSymbol} from ${otherStr}`;
-    }
+    const title = (() => {
+      if (preSwapTransfer) {
+        assert(amount > 0); // foreignCoin can only be involved in receiving ends of swaps
+        return `Accepted $${dollars} from ${otherStr}`;
```
Comment (dcposch) [inline packages/daimo-api/src/server/pushNotifier.ts:249]: so an example push notification is

> **Accepted $42.00 from 0x...**
> You accepted 10000 DEGEN as $42.00 USDC

?

File: packages/daimo-api/src/utils/func.ts [hunk 12]
```diff
@@ -14,3 +14,12 @@ export function memoize<K, I, O>(
...
 }
+
+/**
+ * Returns chunks of size n.
+ * @param {Array<T>} array any array
+ * @param {number} n size of chunk
+ */
+export function* chunks<T>(array: T[], n: number) {
+  for (let i = 0; i < array.length; i += n) yield array.slice(i, i + n);
+}
```
Comment (dcposch) [inline packages/daimo-api/src/utils/func.ts:25]: 👍

File: packages/daimo-userop/src/daimoOpSender.ts [hunk 13]
```diff
@@ -325,6 +328,35 @@ export class DaimoOpSender {
     return this.sendUserOp(op);
   }
 
+  public async executeProposedSwap(
+    swap: ProposedSwap,
+    opMetadata: DaimoOpMetadata
+  ) {
+    console.log(`[OP] execute swap ${swap.fromCoin.token}`);
```
Comment (dcposch) [inline packages/daimo-userop/src/daimoOpSender.ts:335]: print destination & amount too

**Review by dcposch — APPROVED** [review]
🔥 
Looks great to me after comments
The video looks amazing

File: apps/daimo-mobile/src/view/screen/history/HistoryOpScreen.tsx [hunk 14]
```diff
@@ -212,27 +204,32 @@ function TransferBody({
...
       <TouchableOpacity onPress={onShowHelp} hitSlop={8}>
         <TextCenter>
           <TextBodyCaps color={color.grayMid}>{subtitleElems}</TextBodyCaps>
         </TextCenter>
       </TouchableOpacity>
+      {memoText && (
+        <>
+          <Spacer h={16} />
+          <TextCenter>
+            <TextBodyCaps color={color.grayMid}>{memoText}</TextBodyCaps>
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/history/HistoryOpScreen.tsx:229]: Before:

image

After:

image

File: apps/daimo-mobile/src/view/screen/history/shared.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+  daimoChain: DaimoChain,
+  short?: boolean
+) {
+  const chainConfig = env(daimoChain).chainConfig;
+  const coinName = chainConfig.tokenSymbol.toUpperCase();
+
+  if (op.type !== "transfer") return null;
+  if (op.memo) return op.memo;
+  if (op.preSwapTransfer) {
+    if (op.preSwapTransfer.coin.token === "ETH") return null;
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/history/shared.ts:17]: We have no mapping between the origin of the original ETH transfer and the swap, so we can't display anything -- the hack workaround is to just show a labelled Address for the ETH/USDC pool on Base: 

image

File: apps/daimo-mobile/src/view/screen/history/shared.ts [hunk 16]
```diff
@@ -0,0 +1,28 @@
+import { DisplayOpEvent, getForeignCoinDisplayAmount } from "@daimo/common";
+import { DaimoChain } from "@daimo/contract";
+
+import { env } from "../../../logic/env";
+
+export function getSynthesizedMemo(
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/history/shared.ts:6]: > do we support swaps from any non-stablecoins

yes, any coin with a uniswap pool + on the coingecko list.

File: apps/daimo-mobile/src/view/screen/notifications/SwapNotificationRow.tsx [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+        <TextBody color={color.grayMid} style={{ maxWidth: messageWidth }}>
+          Accept {readableAmount} {coin.fullName} from{" "}
+          <TextBody color={color.midnight}>{accName}</TextBody>
+        </TextBody>
+      );
+    } else {
+      return (
+        <TextBody color={color.grayMid} style={{ maxWidth: messageWidth }}>
+          Accept {readableAmount} {coin.fullName} as{" "}
+          <TextBody color={color.midnight}>{homeCoinName}</TextBody>
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/notifications/SwapNotificationRow.tsx:66]: yep exactly

File: packages/daimo-api/src/contract/foreignCoinIndexer.ts (hunk 4 above)
Comment (nalinbhardwaj) [reply packages/daimo-api/src/contract/foreignCoinIndexer.ts:201]: > two userops from the same Daimo account

Yeah this seemed like enough of a future edge case that it seems fine to ignore. Will leave a comment.

File: packages/contract/src/DaimoOffchainUtils.sol [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+contract DaimoOffchainUtils {
+  // Get the ETH balance of a list of addresses in a single call.
+  function batchGetETHBalances(
+    address[] calldata addrs
+  ) external view returns (uint256[] memory) {
+    uint256[] memory ret = new uint256[](addrs.length);
+    for (uint256 i = 0; i < addrs.length; i++) {
+      ret[i] = addrs[i].balance;
+    }
+    return ret;
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply packages/contract/src/DaimoOffchainUtils.sol:14]: I don't think we'd ever need to do this in our codebase for ERC20s? I deployed this on mainnet so will leave for future #1025

File: packages/daimo-api/src/server/pushNotifier.ts (hunk 11 above)
Comment (nalinbhardwaj) [reply packages/daimo-api/src/server/pushNotifier.ts:249]: Yep, see the new tests for the copy:

image

File: packages/daimo-api/package.json [hunk 19]
```diff
@@ -29,10 +29,13 @@
     "@trpc/server": "^11.0.0-next-beta.318",
     "@types/cors": "^2.8.16",
     "@types/geoip-lite": "^1.4.2",
+    "@uniswap/sdk-core": "^4.2.0",
+    "@uniswap/smart-order-router": "^3.27.0",
     "await-lock": "^2.2.2",
     "cors": "^2.8.5",
     "cron": "^3.1.4",
     "dotenv": "^16.4.5",
+    "ethers": "^5.7.2",
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply packages/daimo-api/package.json:38]: We had this in our tree due `userop.js` dependencies fwiw

File: packages/daimo-api/src/contract/ethIndexer.ts (hunk 1 above)
Comment (nalinbhardwaj) [reply packages/daimo-api/src/contract/ethIndexer.ts:120]: yeah I was trying to save the repeated call on previous batch, probably not worth it for code quality.

File: apps/daimo-mobile/src/logic/inAppNotifications.ts [hunk 20]
```diff
@@ -88,6 +97,17 @@ export function useInAppNotifications(): InAppNotificationInfo {
       });
     }
 
+    // Proposed swaps
+    account.proposedSwaps.map((swap) => {
+      console.log(`[IN-APP] pushing swap ${swap.fromCoin}`);
+      notifications.push({
+        type: "swap",
+        marksUnread: true,
+        timestamp: swap.receivedAt || now(),
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/logic/inAppNotifications.ts]: removed, `receivedAt` used to be null for ETH transfers

File: apps/daimo-mobile/src/view/screen/HomeScreen.tsx [hunk 21]
```diff
@@ -256,11 +256,22 @@ function AmountAndButtons({ account }: { account: Account }) {
...
+  const pendingDollars = amountToDollars(
+    account.proposedSwaps.reduce((acc, swap) => acc + swap.toAmount, 0)
+  );
+
   return (
     <TouchableWithoutFeedback accessible={false}>
       <View style={styles.amountAndButtons}>
         <TextLight>Your balance</TextLight>
         <TitleAmount amount={account.lastBalance} />
+        {Number(pendingDollars) > 0.1 && (
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/HomeScreen.tsx]: done

File: apps/daimo-mobile/src/view/screen/history/shared.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/history/shared.ts:10]: added comment

File: apps/daimo-mobile/src/view/sheet/SwapBottomSheet.tsx [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+  );
+}
+
+function SwapInfoRow({ swap }: { swap: ProposedSwap }) {
+  const humanReadableFromAmount = getForeignCoinDisplayAmount(
+    swap.fromAmount,
+    swap.fromCoin
+  );
+
+  const humanReadableToAmount = amountToDollars(swap.toAmount);
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/sheet/SwapBottomSheet.tsx]: done

File: packages/daimo-api/src/contract/ethIndexer.ts [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+import { addrBlockNumKey } from "../utils/indexing";
+import { retryBackoff } from "../utils/retryBackoff";
+
+export type ETHTransfer = {
+  to: Address;
+  value: bigint;
+  blockNumber: number;
+};
+
+export const uniswapETHPoolAddr = "0xd0b53D9277642d899DF5C87A3966A349A798F224";
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply packages/daimo-api/src/contract/ethIndexer.ts]: done

File: packages/daimo-api/src/contract/ethIndexer.ts [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+      try {
+        if (this.isIndexing) {
+          console.log(`[ETHINDEXER] skipping tick, already indexing`);
+          return;
+        }
+        this.isIndexing = true;
+
+        const shovelLatest = await shovel.getShovelLatest();
+
+        // localLatest <= 0 when there are no new blocks in shovel
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply packages/daimo-api/src/contract/ethIndexer.ts]: done

File: packages/daimo-api/src/contract/ethIndexer.ts [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+            this.vc.publicClient.readContract({
+              abi: daimoOffchainUtilsABI,
+              address: daimoOffchainUtilsAddress,
+              functionName: "batchGetETHBalances",
+              args: [batch],
+              blockNumber: BigInt(blockNum),
+            })
+          );
+        }
+      })();
```
[the remark this answers is not in the document]
Comment (nalinbhardwaj) [reply packages/daimo-api/src/contract/ethIndexer.ts]: done, changed a bit more

File: packages/daimo-api/src/contract/ethIndexer.ts (hunk 2 above)
Comment (nalinbhardwaj) [reply packages/daimo-api/src/contract/ethIndexer.ts]: done

File: packages/daimo-api/src/contract/foreignCoinIndexer.ts (hunk 3 above)
Comment (nalinbhardwaj) [reply packages/daimo-api/src/contract/foreignCoinIndexer.ts]: moved to shovel/watcher init

File: packages/daimo-api/src/network/uniswapClient.ts (hunk 5 above)
Comment (nalinbhardwaj) [reply packages/daimo-api/src/network/uniswapClient.ts:31]: done

File: packages/daimo-api/src/network/uniswapClient.ts (hunk 6 above)
Comment (nalinbhardwaj) [reply packages/daimo-api/src/network/uniswapClient.ts]: reduced to 10min

File: packages/daimo-api/src/network/uniswapClient.ts (hunk 7 above)
Comment (nalinbhardwaj) [reply packages/daimo-api/src/network/uniswapClient.ts:153]: done

File: packages/daimo-api/src/server/coinList.ts (hunk 8 above)
Comment (nalinbhardwaj) [reply packages/daimo-api/src/server/coinList.ts]: done

File: packages/daimo-userop/src/daimoOpSender.ts (hunk 13 above)
Comment (nalinbhardwaj) [reply packages/daimo-userop/src/daimoOpSender.ts]: done

**Pull request merged**